From 528768621ebb8417345560e2650961528092bad3 Mon Sep 17 00:00:00 2001
From: Jiri Pirko <jpirko@redhat.com>
Date: Mon, 14 Jul 2014 11:47:58 -0400
Subject: [net] gre: add link local route when local addr is any

Message-id: <1405338479-17101-2-git-send-email-jpirko@redhat.com>
Patchwork-id: 85734
O-Subject: [patch rhel7 net 2/3] gre: add link local route when local addr is any
Bugzilla: 1107369
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>

BZ1107369
https://brewweb.devel.redhat.com/taskinfo?taskID=7696184

upstream commit 08b44656c08c8c2f73cdac2a058be2880e3361f2
Author: Nicolas Dichtel <nicolas.dichtel@6wind.com>
Date:   Mon Feb 17 14:22:21 2014 +0100

    gre: add link local route when local addr is any

    This bug was reported by Steinar H. Gunderson and was introduced by commit
    f7cb8886335d ("sit/gre6: don't try to add the same route two times").

    root@morgental:~# ip tunnel add foo mode gre remote 1.2.3.4 ttl 64
    root@morgental:~# ip link set foo up mtu 1468
    root@morgental:~# ip -6 route show dev foo
    fe80::/64  proto kernel  metric 256

    but after the above commit, no such route shows up.

    There is no link local route because dev->dev_addr is 0 (because local ipv4
    address is 0), hence no link local address is configured.

    In this scenario, the link local address is added manually: 'ip -6 addr add
    fe80::1 dev foo' and because prefix is /128, no link local route is added by the
    kernel.

    Even if the right things to do is to add the link local address with a /64
    prefix, we need to restore the previous behavior to avoid breaking userpace.

Signed-off-by: Jiri Pirko <jpirko@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv6/addrconf.c b/net/ipv6/addrconf.c
index 1b61479..f068408 100644
--- a/net/ipv6/addrconf.c
+++ b/net/ipv6/addrconf.c
@@ -2763,6 +2763,8 @@ static void addrconf_gre_config(struct net_device *dev)
  ipv6_addr_set(&addr,  htonl(0xFE800000), 0, 0, 0);
  if (!ipv6_generate_eui64(addr.s6_addr + 8, dev))
   addrconf_add_linklocal(idev, &addr);
+ else
+  addrconf_prefix_route(&addr, 64, dev, 0, 0);
 }
 #endif
 
-- 
1.7.1